### PR TITLE
Allow persisting events when recovery has completed

### PR DESCRIPTION
**Pull request by ortibazar**
Refs #21736

### CODE REVIEW

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala [hunk 1]
```diff
@@ -577,6 +576,17 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
...
       }
+
+      private def transitToProcessingState(): Unit = {
+        if (eventBatch.nonEmpty) flushBatch()
+
+        if (pendingStashingPersistInvocations > 0) changeState(persistingEvents)
+        else {
+          changeState(processingCommands)
+          internalStash.unstashAll()
+        }
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: ah, that makes sense

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala [hunk 2]
```diff
@@ -646,6 +646,14 @@ object PersistentActorSpec {
     }
   }
 
+  class PersistInRecovery(name: String) extends ExamplePersistentActor(name) {
+    override def receiveRecover = super.receiveRecover orElse {
+      case RecoveryCompleted ⇒
+        persist(Evt(RecoveryCompleted))(updateState)
```
Comment (patriknw) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala]: what happens if `persist` is done when processing events in `receiveRecover`, i.e. not only when `RecoveryCompleted`?

Comment (patriknw) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala]: would also be interesting with a test that is using `persistAsync` from `receiveRecover`

File: akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala [hunk 3]
```diff
@@ -43,7 +43,7 @@ trait AsyncWriteJournal extends Actor with WriteJournalBase with AsyncRecovery {
       case "fail"                  ⇒ ReplayFilter.Fail
       case "warn"                  ⇒ ReplayFilter.Warn
       case other ⇒ throw new IllegalArgumentException(
-        s"invalid replay-filter.mode [$other], supported values [off, repair, fail, warn]")
+        s"invalid replay-filter.mode [$other], supported values [off, repair-by-discard-old, fail, warn]")
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala]: good catch

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
OK TO TEST

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala [hunk 4]
```diff
@@ -316,6 +316,7 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
    * @param handler handler for each persisted `event`
    */
   def persist[A](event: A)(handler: A ⇒ Unit): Unit = {
+    if (recoveryRunning) throw new IllegalStateException("Cannot persist event during recovery")
```
Comment (drewhk) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: isn't the whole point to be able to persist events during recovery? I am confused here.

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala [hunk 5]
```diff
@@ -1119,6 +1136,20 @@ abstract class PersistentActorSpec(config: Config) extends PersistenceSpec(confi
...
+      persistentActor ! GetState
+      expectMsg(List("a-1", "a-2", "rc-1", "rc-2"))
+      persistentActor ! GetState
+      expectMsg(List("a-1", "a-2", "rc-1", "rc-2", "rc-3"))
+      persistentActor ! Cmd("invalid")
+      persistentActor ! GetState
+      expectMsg(List("a-1", "a-2", "rc-1", "rc-2", "rc-3", "invalid"))
+      watch(persistentActor)
+      persistentActor ! "boom"
+      expectTerminated(persistentActor)
```
Comment (johanandren) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala]: So it is actually fail-fast on persist during recovery, but allow it in `RecoveryCompleted`?

Comment (patriknw) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala]: yes

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala (hunk 4 above)
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: Cannot persist during replay. Events can be persisted when receiving RecoveryCompleted or later.

**Comment by ktoso** [comment]
LGTM, thanks - nice improvement!

**Pull request merged**